drivers/adc/adc_max32.c

```diff
@@ -0,0 +1,341 @@
+{
+	struct max32_adc_data *data = CONTAINER_OF(ctx, struct max32_adc_data, ctx);
+
+	data->sample_channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	adc_max32_start_channel(data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	struct max32_adc_data *data = CONTAINER_OF(ctx, struct max32_adc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int start_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	struct max32_adc_data *data = dev->data;
+	uint32_t num_of_sample_channels = POPCOUNT(seq->channels);
+	uint32_t num_of_sample = 1;
+	int ret = 0;
+
+	if (seq->resolution != data->resolution) {
+		LOG_ERR("Unsupported resolution (%d)", seq->resolution);
+		return -ENOTSUP;
+	}
+	if (seq->channels == 0) {
+		return -EINVAL;
+	}
+	if ((data->channels & seq->channels) != seq->channels) {
+		return -EINVAL;
+	}
+
+	ret = Wrap_MXC_ADC_AverageConfig(seq->oversampling);
+	if (ret != 0) {
+		return -EINVAL;
+	}
+
+	if (seq->options) {
+		num_of_sample += seq->options->extra_samplings;
+	}
+	if (seq->buffer_size < (num_of_sample * num_of_sample_channels)) { /* Buffer size control */
+		return -ENOMEM;
+	}
+
+	data->buffer = seq->buffer;
+	adc_context_start_read(&data->ctx, seq);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_max32_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	struct max32_adc_data *data = dev->data;
+	int ret;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	ret = start_read(dev, seq);
+	adc_context_release(&data->ctx, ret);
+
+	return ret;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_max32_read_async(const struct device *dev, const struct adc_sequence *seq,
+				struct k_poll_signal *async)
+{
+	struct max32_adc_data *data = dev->data;
+	int ret;
+
+	adc_context_lock(&data->ctx, true, async);
+	ret = start_read(dev, seq);
+	adc_context_release(&data->ctx, ret);
+
+	return ret;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static int adc_max32_channel_setup(const struct device *dev, const struct adc_channel_cfg *cfg)
+{
+	const struct max32_adc_config *conf = dev->config;
+	struct max32_adc_data *data = dev->data;
+	wrap_mxc_adc_scale_t wrap_mxc_scale;
+	uint8_t adc_reference;
+	int ret = 0;
+
+	if (cfg->channel_id >= conf->channel_count) {
+		LOG_ERR("Invalid channel (%u)", cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Invalid channel acquisition time");
+		return -EINVAL;
+	}
+
+	if (cfg->differential) {
+		LOG_ERR("Differential sampling not supported");
+		return -ENOTSUP;
+	}
+
+	switch (cfg->reference) {
+	case ADC_REF_INTERNAL:
+		adc_reference = ADI_MAX32_ADC_REF_INTERNAL;
+		break;
+	case ADC_REF_VDD_1_2:
+		adc_reference = ADI_MAX32_ADC_REF_VDD_1_2;
+		break;
+	case ADC_REF_EXTERNAL0:
+		adc_reference = ADI_MAX32_ADC_REF_EXT0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
```

Comment:
Is 2.048V not supported?

Reply:
There is only 1 internal reference voltage option in Zephyr.
![image](https://github.com/user-attachments/assets/f19b2200-6d8b-4ef3-b293-bd6763817183)

MCUs support external voltage from 2.048V to VDDA.
![image](https://github.com/user-attachments/assets/f35a38b7-1514-44bd-8610-6063665d5548)

Because of that, I choose 1.25V internal voltage option for Zephyr's internal voltage option.